rpc/rpcserver/server.go

```diff
@@ -2314,3 +2328,125 @@ func (s *messageVerificationServer) VerifyMessage(ctx context.Context, req *pb.V
+
+		inputs[i] = &pb.DecodedTransaction_Input{
+			PreviousTransactionHash:  txIn.PreviousOutPoint.Hash[:],
+			PreviousTransactionIndex: txIn.PreviousOutPoint.Index,
+			Tree:               pb.DecodedTransaction_Input_TreeType(txIn.PreviousOutPoint.Tree),
+			Sequence:           txIn.Sequence,
+			AmountIn:           txIn.ValueIn,
+			BlockHeight:        txIn.BlockHeight,
+			BlockIndex:         txIn.BlockIndex,
+			SignatureScript:    txIn.SignatureScript[:],
+			SignatureScriptAsm: disbuf,
+		}
+	}
+
+	return inputs
+}
+
+func marshalDecodedTxOutputs(mtx *wire.MsgTx, chainParams *chaincfg.Params) []*pb.DecodedTransaction_Output {
+	outputs := make([]*pb.DecodedTransaction_Output, len(mtx.TxOut))
+	txType := stake.DetermineTxType(mtx)
+
+	for i, v := range mtx.TxOut {
+		// The disassembled string will contain [error] inline if the
+		// script doesn't fully parse, so ignore the error here.
+		disbuf, _ := txscript.DisasmString(v.PkScript)
+
+		// Attempt to extract addresses from the public key script.  In
+		// the case of stake submission transactions, the odd outputs
+		// contain a commitment address, so detect that case
+		// accordingly.
+		var addrs []dcrutil.Address
+		var encodedAddrs []string
+		var scriptClass txscript.ScriptClass
+		var reqSigs int
+		var commitAmt *dcrutil.Amount
+		if txType == stake.TxTypeSStx && (i%2 != 0) {
+			scriptClass = txscript.StakeSubmissionTy
+			addr, err := stake.AddrFromSStxPkScrCommitment(v.PkScript,
+				chainParams)
+			if err != nil {
+				encodedAddrs = []string{fmt.Sprintf(
+					"[error] failed to decode ticket "+
+						"commitment addr output for tx hash "+
+						"%v, output idx %v", mtx.TxHash(), i)}
+			} else {
+				encodedAddrs = []string{addr.EncodeAddress()}
+			}
+			amt, err := stake.AmountFromSStxPkScrCommitment(v.PkScript)
+			if err != nil {
+				commitAmt = &amt
+			}
+		} else {
+			// Ignore the error here since an error means the script
+			// couldn't parse and there is no additional information
+			// about it anyways.
+			scriptClass, addrs, reqSigs, _ = txscript.ExtractPkScriptAddrs(
+				v.Version, v.PkScript, chainParams)
+			encodedAddrs := make([]string, len(addrs))
+			for j, addr := range addrs {
+				encodedAddrs[j] = addr.EncodeAddress()
+			}
+		}
+
+		outputs[i] = &pb.DecodedTransaction_Output{
+			N:                  uint32(i),
+			Value:              v.Value,
+			Version:            int32(v.Version),
+			Addresses:          encodedAddrs,
+			Script:             v.PkScript,
+			ScriptAsm:          disbuf,
+			ScriptClass:        pb.DecodedTransaction_Output_ScriptClass(scriptClass),
+			RequiredSignatures: int32(reqSigs),
+		}
+		if commitAmt != nil {
+			outputs[i].CommitmentAmount = int64(*commitAmt)
+		}
+	}
+
+	return outputs
+}
+
+func (s *decodeMessageServer) DecodeRawTransaction(ctx context.Context, req *pb.DecodeRawTransactionRequest) (
+	*pb.DecodeRawTransactionResponse, error) {
+
+	serializedTx := req.GetSerializedTransaction()
```

Comment:
we don't need to use the getter functions at all, can just use req.SerializedTransaction directly